Language Server Protocol

ellerh opened the pull request:
As an experiment, I wrote a LSP server for Forth.

The server does very little: it scans the files and buffers in a
project for definitions and records them so that M-. can find them.

It took quite a bit of code (and time) to get it this far.  Are you
interested to make this a part of forth-mode or would you prefer that I
keep this as a separate project?

On the client side, I used eglot. So there is very little new elisp
code.

ellerh committed:
Experimental Forth Language Server.

larsbrinkhoff commented:
Thank you, this seems interesting.  I don't have time now, but I will take a look soon.

Regarding separation, if it's potentially useful elsewhere I think a separate repository might make sense.

ellerh committed:
Delete fd-stream.fth. An old file that accidentally was added to the repo.

ellerh committed:
Wait for the "exit" notification before terminating.
That was apparently the problem that eglot-shutdown encountered previously.

ellerh committed:
Update test/lsp.fth for modified shutdown behaviour.

ellerh committed:
Truncate overly long tokens.
Without this, we crash while indexing the openfirmware repo.

ellerh committed:
When indexing, delete existing definitions only for textdocs.

This is a workaround for the slow db-delete-definitions-with-uri.  By
doing this, the time to index the openfirmware repo dropped from
unpractical 25 seconds to more reasonable 3 seconds.

One day we may have better data structures in the db and do it
properly again.

ellerh committed:
Fix typo.

ellerh committed:
Fix bug in JB-ADD-CODEPOINT.

It's unsafe to use the address returned from REGION-NEXT-FREE after
REGION-BLANK.  Introduce a hopefully less error prone REGION-ADD.

This is, unfortunately, quite hard to test.

ellerh committed:
Recognize a few more defining words.

ellerh committed:
Delete existing definitions when a buffer gets closed.

ellerh committed:
Add some tests for various defining words.

ellerh committed:
Reimplement reader-read-line-crlf.
Hopefully less confusing this time.

ellerh committed:
Cosmetics: use BEGIN-STRUCTURE/END-STRUCTURE more consistently

ellerh committed:
More struct cosmetics.

ellerh committed:
Parse and record stack comments for colon definitions.

ellerh committed:
Add support for hover requests.

ellerh committed:
Merge remote-tracking branch 'lars/master' into fools

ellerh committed:
Merge remote-tracking branch 'lars/master' into fools

ellerh committed:
Make the stack effect of all standard words known

* std-effects.txt: New file.
* lsp.fth (%insert-std-effects, %bl?): New.
* db.fth (%db-uri-match?): Allow empty URLs.
